File: Codes/Bingran Li/Tree/102. Binary Tree Level Order Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
      
        // someone's solution: very good-----try something new next question
        ArrayList<List<Integer>> finalList = new ArrayList<List<Integer>> ();
        LinkedList <TreeNode> myQue = new LinkedList <TreeNode>(); // used to be integer
        if(root==null)
            return finalList;
        myQue.add(root);
        while(myQue.size()>0){
            ArrayList<Integer> temList = new ArrayList<Integer>();
            int size =myQue.size();
            for(int i=0;i<size;i++){  //----------> myQue.size() change over time!!!
                TreeNode myNode = myQue.poll();
                
                temList.add(myNode.val);
                if(myNode.left!=null) myQue.add(myNode.left);
                if(myNode.right!=null) myQue.add(myNode.right);
                
            }
            
            finalList.add(temList);
        }
        return finalList;
        
        
        
        
        
    }
}

File: Codes/Kan Zheng/C++/Tree/102. Binary Tree Level Order Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> vecvec;
        vector<int> vec;
        if(root==NULL)
            return vecvec;
        //1:conbine left and right
        vecvec = conbine_vv(levelOrder(root->left),levelOrder(root->right));
        //2:add it self
        vec.push_back(root->val);
        vecvec.insert(vecvec.begin(),vec);
        //return vecvec
        return vecvec;
    }
    vector<vector<int>> conbine_vv(vector<vector<int>> left,vector<vector<int>> right)
    {
        vector<vector<int>> vec;
        if(left.empty())
            return right;
        else if(right.empty())
            return left;
        bool useleft;
        if(left.size()>right.size())
            {useleft=true;vec=left;}
        else
            {useleft=false;vec=right;}
        int i,j;
        for(i=0,j=0;i<left.size()&&j<right.size();i++,j++)
        {
            if(useleft)
                vec[i]=conbine_v(left[i],right[j]);
            else
                vec[j]=conbine_v(left[i],right[j]);
        }
        return vec;
    }
    //conbine a and b
    vector<int> conbine_v(vector<int> a,vector<int> b)
    {
        vector<int> vec=a;
        for(int i=0;i<b.size();i++)
            vec.push_back(b[i]);
        return vec;
    }
};

//use queue
